core/domain/user_services.py

```diff
@@ -2287,3 +2291,272 @@ def get_dashboard_stats(user_id):
+                furthest_reached_checkpoint_in_current_exploration) <= (
+                    checkpoints_in_current_exploration.index(state_name))):
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_exp_version
+                ) = exp_version
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_state_name
+                ) = state_name
+
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = (
+        exp_version)
+    exploration_user_model.most_recently_reached_checkpoint_state_name = (
+        state_name)
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def set_user_has_viewed_lesson_info_modal_once(user_id):
+    """Updates the user's settings once he has viewed the lesson info modal.
+
+    Args:
+        user_id: str. The Id of the user.
+    """
+
+    user_settings = get_user_settings(user_id)
+    user_settings.mark_lesson_info_modal_viewed()
+    _save_user_settings(user_settings)
+
+
+def update_learner_checkpoint_progress_on_restart(user_id, exploration_id):
+    """Sets the most recently reached checkpoint on restart exploration event.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        # This case normally won't arise since as soon as the exploration
+        # starts most_recently_reached_checkpoint is updated (First state of
+        # an exploration is always a checkpoint). This is just a double check
+        # in case learner hits the restart button while the other content of
+        # the page is loading.
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = None
+    exploration_user_model.most_recently_reached_checkpoint_state_name = None
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def sync_learner_checkpoint_progress_with_current_exp_version(
+    user_id,
+    exploration_id):
+    """Synchronizes the most recently reached checkpoint and the furthest
+    reached checkpoint with the latest exploration.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+
+    Returns:
+        ExplorationUserData. The domain object corresponding to the given user
+        and exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+
+    if exploration_user_model is not None:
+        latest_exploration = exp_fetchers.get_exploration_by_id(exploration_id)
+        most_recently_interacted_exploration = (
```

Comment:
One question. It is actually possible for learners to play an older version of an exploration by sticking `?v=...` to the end of the URL. This use case happens, but is rare (it tends to happen in e.g. iframes which pin a particular exploration version). 

Shall we just not do any checkpoint-related syncing if that case arises? Might need to check on this with the product team though.

Reply:
After discussion with Diana and Pankaj, we are now marking user checkpoint progress as None for the particular GET request when an older version of an exploration is tried to be fetched. As a result, for such requests, the exploration shall begin from the first state. The other changes where we do not allow users to save progress in such scenarios are to be made in the frontend with the link pointing to where they can access the most up to date version of the lesson.

Reply:
Sounds reasonable. But where do you make these changes? I would have expected to see them in reader.py, as well as a backend test for them.

Reply:
In reader.py, I am updating values for 'furthest_reached_checkpoint' and 'most_recently_reached_checkpoint' only when the exploration_user_data is not None and version asked in GET request in None. [link to the check](https://github.com/oppia/oppia/pull/15213/files#diff-a82c33370255276ab4a04c32203a00a31dd3a53ee96bab27e1eedf0e9a7b39cfR297).

I'll write necessary tests and explanatory comment right away.

Reply:
Done!